tiledb/sm/global_state/tbb_state.cc

```diff
@@ -0,0 +1,108 @@
+/**
+ * @file   tbb_state.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2018 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ This file declares the Intel TBB threading scheduler state, if TileDB is built
+ with Intel TBB.
+ */
+
+#include "tiledb/sm/global_state/tbb_state.h"
+
+#ifdef HAVE_TBB
+
+#include <tbb/task_scheduler_init.h>
+#include <sstream>
+
+namespace tiledb {
+namespace sm {
+namespace global_state {
+
+/** The TBB scheduler, used for controlling the number of TBB threads. */
+static std::unique_ptr<tbb::task_scheduler_init> tbb_scheduler_;
+
+/** The number of TBB threads the scheduler was configured with **/
+static int tbb_nthreads_;
+
+Status init_tbb(tiledb::sm::Config* config) {
+  int nthreads = config ? config->sm_params().num_tbb_threads_ :
+                          tbb::task_scheduler_init::automatic;
```

Comment:
Should probably use `constants::num_tbb_threads` here since it's already defined.